### PR TITLE
chore: log through Fluent Bit on Kubernetes [DET-4622]

### PR DESCRIPTION
**Pull request by dzhu**
## Description

This makes logging on Kubernetes go through Fluent Bit by creating a
sidecar container running Fluent Bit for each trial pod that shares a
volume with the main container. Logs go to files in the volume and are
picked up by Fluent Bit's file tailing input. In order to limit disk
usage, logs are piped, via shell shenanigans in order to get stdout and
stderr treated separately, through a tool called multilog, which handles
rotating output files based on size. (As opposed to using another
container running logrotate or similar, doing it directly on the output
side avoids messing around with truncating files in place and the like.)

The logs volume is configured to be backed by memory, so logs don't have
to hit a disk just to get from one container to another.

## Test Plan

- [x] check that logs come through correctly when everything is run on: (1) agents with PostgreSQL logs (2) agents with Elastic logs (3) Kubernetes with PostgreSQL logs [all including TLS and non-TLS for the master and Elastic]
- [ ] check that dtrain works with the extra indirection (I tried something in a shell locally that should be equivalent and it worked as expected, but we'd want to run dtrain for real to be sure)

### CODE REVIEW

File: master/static/srv/entrypoint.sh [hunk 1]
```diff
@@ -3,12 +3,31 @@
...
 # streams of this process, allowing distributed training logs to route through
 # individual containers rather than all going through SSH back to agent 0.
-ln -sf /proc/$$/fd/1 /run/determined/train/stdout.log
-ln -sf /proc/$$/fd/2 /run/determined/train/stderr.log
+ln -sf /proc/$$/fd/1 "$STDOUT_FILE"
+ln -sf /proc/$$/fd/2 "$STDERR_FILE"
+
+if [ -n "$DET_K8S_LOG_TO_FILE" ]; then
+    # TODO: Do this in the base environment image.
+    apt-get update && apt-get install -y daemontools
```
Comment (stoksc) [inline master/static/srv/entrypoint.sh:20]: For airgapped k8s clusters (maybe there is one 🤷), we definitely should move this before landing.

**Review by stoksc — COMMENTED** [review]
this looks great

File: master/static/srv/entrypoint.sh (hunk 1 above)
Comment (dzhu) [reply master/static/srv/entrypoint.sh:20]: Oh yeah, this is totally temporary for the PR. Just didn't feel like waiting on a cycle with the environments repo to get the first version out, though I guess I can go ahead and get that started at any point.

**Review by stoksc — COMMENTED** [review]
lgtm, just had some questions.

File: master/pkg/fluent/fluent.go [hunk 2]
```diff
@@ -0,0 +1,227 @@
+package fluent
```
Comment (stoksc) [inline master/pkg/fluent/fluent.go:1]: We might could add a least one unit tests that checks a fully saturated fluent config is formed correctly

File: master/internal/kubernetes/pod.go [hunk 3]
```diff
@@ -71,6 +74,9 @@ func newPod(
 	namespace string,
 	masterIP string,
 	masterPort int32,
+	masterTLSConfig model.TLSClientConfig,
+	loggingTLSConfig model.TLSClientConfig,
+	loggingConfig model.LoggingConfig,
 	podInterface typedV1.PodInterface,
 	configMapInterface typedV1.ConfigMapInterface,
 	resourceRequestQueue *actor.Ref,
```
Comment (stoksc) [inline master/internal/kubernetes/pod.go:82]: @aaron276h mentioned this but it's a good point, we probably want to add the fluent container to the `containerNames` map below so when we do healthchecks we also terminate on its failures (maybe run this behavior by @vishnu2kmohan, do you want to fail on log failures or keep running without logs?)

**Review by stoksc — APPROVED** [review]
lgtm, nice figuring all this k8s/fluent stuff out

File: master/internal/kubernetes/pod.go (hunk 3 above)
Comment (aaron276h) [reply master/internal/kubernetes/pod.go:82]: One more thing to keep in mind here, but unfortunately it's only available is k8s 1.18+ is explicitly labeling it a sidecar container which resolves issues related to startup ordering. 

It is also worth checking some failure scenarios, to make sure we exit properly and don't get stuck in an infinite loop.

Comment (vishnu2kmohan) [reply master/internal/kubernetes/pod.go:82]: > when we do healthchecks

Is adding health-checks for the fluent container not in-plan as things stand? It's good practice to enable health-checks for all containers (esp. dependencies), unless there are some good reasons why we don't already have this, today.

> do you want to fail on log failures or keep running without logs?

Ideally we'd fail if we're unable to log. 

@stoksc Are you referring to a startup failure or runtime failure? Can you share a bit more detail on the (expected) failure conditions we need to handle?

Comment (stoksc) [reply master/internal/kubernetes/pod.go:82]: > when we do healthchecks

I mean when the code does them (we already have them), as is we just never restart and if the container is terminated we assume it has failed, right @aaron276h?

> do you want to fail on log failures or keep running without logs?

Ok, we should do what I'm suggesting then, along with the testing @aaron276h is suggesting.

> One more thing to keep in mind here, but unfortunately it's only available is k8s 1.18+ is explicitly labeling it a sidecar container which resolves issues related to startup ordering.

Interesting, hadn't heard of this, only knew of normal and init containers.

Comment (dzhu) [reply master/internal/kubernetes/pod.go:82]: Good catch; I added the name to the map for trial containers and tested that the Fluent Bit container failing (either on startup or by being killed later) kills the trial without any issues. Messaging could be better when that happens, but it works.

Did the sidecar thing actually get into Kubernetes? Based on https://github.com/kubernetes/kubernetes/pull/79649 and https://github.com/kubernetes/kubernetes/pull/80744, it looks like it actually got put off, unfortunately.

**Pull request merged**